### ISSUE TITLE
Discrepancy between np.take and jnp.take

### ISSUE THREAD

**Issue by gianlucadetommaso** [google/jax#10967]
There seem to be an unexpected discrepancy between the behaviour of take in `jax.numpy.take` and `numpy.take` for indices out-of-bound. Here is a minimal example.

`numpy.take(np.ones(2), np.array([1, 2]))` gives the error message `IndexError: index 2 is out of bounds for axis 0 with size 2`, while `jax.numpy.take(np.ones(2), np.array([1, 2]))` returns `DeviceArray([1., 1.], dtype=float32)`.

The behaviour of `jax.numpy.take` looks wrong to me. Thoughts?

**Comment by sokrypton** [comment]
interesting... I wanted to see exactly what it was returning when value is unavailable. 

<img width="746" alt="image" src="https://user-images.githubusercontent.com/4187522/171880316-b53f4fbe-d328-40ec-9e71-b9efc95d7ef2.png">

**Comment by sokrypton** [comment]
I'm guessing the reason for this is because at jit compile time it's not known what indices will be available, as these can dynamically change.

**Comment by jakevdp** [comment]
This is a known behavior in JAX; you can read about it here: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#out-of-bounds-indexing

**Comment by jakevdp** [comment]
With some indexing constructs in JAX, you can control the out-of-bounds behavior to a degree; see https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html

**Comment by YouJiacheng** [comment]
@jakevdp Could we validate dynamic value inside transformed function, e.g. `vmap` and `grad`?(Not during tracing, but during interpreting).
If we can do so, it might be good to validate dynamic value everywhere except inside jitted function as a debugging functionality. (might be controlled by a flag)
(But considering the significantly inferior performance of un-jitted JAX, it might be useless to debug without jit😭.)

**Comment by jakevdp** [comment]
Yeah, it would be possible to validate indices for concrete avals within autodiff, but I don't think that's the preferred solution because it introduces an asymmetry between JIT and non-JIT code. I think that `checkify` is a better route to adding this kind of validation, and I think @mattjj & @LenaMartens have been thinking about this a bit.

**Comment by YouJiacheng** [comment]
Thanks @jakevdp ‘s response, I considered the consistency between JIT and non-JIT code too. I agree that isn't the preferred solution. And thanks for letting me know `checkify` can do this kind of validation!
